docs: add maven-use reference page (#1179)
canonical/craft-parts: docs/common/craft-parts/reference/plugins/maven_use_plugin.rst

.. _craft_parts_maven_use_plugin:

Maven Use plugin
================

The Maven Use plugin allows for setting up a `Maven repository`_ for `Maven`_-based
projects. It is a companion plugin meant to be used with the
:ref:`craft_parts_maven_plugin`.

Keys
----

This plugin has no unique keys.

.. _maven_use_self-contained_start:

Attributes
----------

This plugin supports the ``self-contained`` build attribute. By declaring this
attribute on all parts using the Maven or Maven Use plugin, Maven will only use

> locally-available projects as dependencies, creating an offline build of that artifact.

Author: The "self-contained" bit is pretty nuanced, so some specific feedback could be useful here. What I need to capture is the fact that using the build-attribute does not mean your invokation of `somecraft pack` will not connect to the internet whatsoever. If you have a part with `source: https://github.com/...` for example, it will still download that part's source from GitHub. What the build attribute actually does is make it so that Maven itself will be unable to connect to the internet, instead being forced to only use locally available artifacts. This is why I made the suggestion somewhere else in the document that you can make a build dependency available by building it in a previous `maven-use` part.

> This plugin supports the ``self-contained`` build attribute. By declaring this
> attribute on all parts using the Maven or Maven Use plugin, Maven will only use
> locally-available projects as dependencies, creating an offline build of that artifact.

Reviewer: Ah, I see. In that case, maybe we could do something like this?

```suggestion
This plugin supports the ``self-contained`` build attribute. When declared on all parts
using the Maven or Maven Use plugin, this attribute restricts Maven builds to the local
build environment by sourcing their dependencies locally.
```

With the ``self-contained`` attribute, Maven Use may additionally rewrite the version
specification of any project dependencies based on what is actually available on-disk.
This behavior can be avoided by having the exact version already present at build time,
such as by building it in a previous part with the Maven Use plugin.

.. _maven_use_self-contained_end:

.. _maven_use_details_begin:

Dependencies
------------

The Maven plugin needs the ``mvn`` executable to build Maven projects but does not

> provision it by itself, to allow flexibility in the choice of compiler version.

Reviewer: Adding this separately because it would be inconsistent with the rest of the plugin docs

```suggestion
provision it to allow flexibility in the choice of version.
```

You don't have to do anything with this. I've just never liked the flow of "it by itself" :o

To provide ``mvn``, one can either specify the ``maven`` Ubuntu package as a
``build-package``, or define a ``maven-deps`` part. In the case of the latter, all
parts using Maven should declare that they come after the ``maven-deps`` part. In this
case, the plugin will assume that this new part will stage the ``maven`` executable to
be used in the build step. This can be useful, for example, in cases where a specific,
unreleased version of Maven is desired but unavailable as a snap or Ubuntu package.

.. _maven_use_details_end:

How it works
------------

During the build step the plugin performs the following actions:

* Create a Maven settings file that configures proxy settings, points to the local
  Maven repository created by Craft Parts, and optionally disables network connections
  when using the ``self-contained`` build attribute.
* Updates the :file:`pom.xml` file for project from the ``source`` key to tell Maven to

>   deploy the final artifact to the local repository.

Author: That's much better - though I'm curious, what's our stance on optional plurals (e.g. "file(s)")? Technically both "pom.xml file" and "artifact" could, situationally, be plural.

Maven Use plugin
================

> The Maven Use plugin packages a `Maven`_-based project for use by other plugins using
> the Maven Use plugin or the :ref:`craft_parts_maven_plugin`. It differs from the Maven
> plugin in that it deploys the artifact to an internal `Maven repository`_ instead of
> placing it in the final output.

Reviewer: ```suggestion
The Maven Use plugin packages `Maven`_ projects and, unlike the
:ref:`craft_parts_maven_plugin`, deploys the artifacts to an internal `Maven
repository`_ . From this repository, the artifacts can be accessed by any other parts
using the Maven or Maven Use plugins.
```

Attributes
----------

> This plugin supports the ``self-contained`` build attribute. Using this attribute will
> disable the default `Maven Central repository`_ for every dependency including plugins.
> All necessary dependencies will then need to be provided either as build packages or in
> earlier parts.

Reviewer: ```suggestion
This plugin supports the ``self-contained`` build attribute. Declaring this attribute
prevents access to the default `Maven Central Repository`_. All dependencies, including
those for plugins, must then be provided as build packages or in an earlier part.
```

> When using this attribute, Maven Use may rewrite the version specification

Reviewer: ```suggestion
When this attribute is declared, Maven Use may rewrite the version specification
```

Examples
--------

> The following snippet declares a part named ``java-dep`` using the ``maven-use`` plugin
> and a ``java-main`` part. The ``java-main`` part depends on the contents of
> ``java-dep`` to build. Correct ordering is achieved with the use of the ``after`` key in
> the ``java-main`` part.

Reviewer: ```suggestion
The following snippet declares two parts: ``java-dep``, which uses the ``maven-use``
plugin, and ``java-main``. Before ``java-main`` can build, the contents of ``java-dep``
must be staged. This dependency is handled by declaring that ``java-main`` must
build ``after`` the ``java-dep`` part.
```

.. code-block::

    parts:
      java-dep:
        source: dep/
        plugin: maven-use
        build-packages:
          - maven
      java-main:
        source: main/
        plugin: maven
        build-packages:
          - maven
        after:
          - java-dep

> The following snippet declares a part named ``java-jacoco`` using the ``maven-use``

Reviewer: ```suggestion
The following snippet declares two parts: ``java-jacoco``, which uses the ``maven-use``
```

> plugin and a ``java-main`` part. This build is done using the ``self-contained``
> build attribute to control what Maven downloads for the build. The ``java-main`` part
> declares a dependency on the ``jacoco-maven-plugin`` in its :file:`pom.xml` file. As
> with the previous example, correct ordering is achieved with the use of the ``after``
> key in the ``java-main`` part.

Reviewer: ```suggestion
plugin, and ``java-main``. To restrict access to the Maven Central Repository, both
parts declare the ``self-contained`` build attribute. The :file:`pom.xml` file of
``java-main`` declares ``java-jacoco`` as a dependency, which is handled by declaring
that ``java-main`` must build ``after`` the ``java-jacoco`` part.
```
